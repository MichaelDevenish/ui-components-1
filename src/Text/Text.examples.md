#### Basic Text

```jsx
  <Text>
    Hello! I am text
  </Text>
```

#### Skeleton Text

```jsx
  <Text
    skeletonOptions={{
      showSkeleton: true,
      shape: 'line',
      width: 200
    }}
  >
    Hello! I am text
  </Text>
```

#### Heavy

```jsx
  <Text heavy>
    Hello! I am text
  </Text>
```

#### Subtle

```jsx
  <Text subtle>
    Hello! I am text
  </Text>
```

#### Upper

```jsx
  <Text upper>
    Hello! I am text
  </Text>
```

#### Small

```jsx
  <Text size='small'>
    Hello! I am text
  </Text>
```

#### Large

```jsx
  <Text size='large'>
    Hello! I am text
  </Text>
```

#### Inline (false)

```jsx
<div>
  <Text inline={false}>
    Hello!
  </Text>
  <Text>
    I am text
  </Text>
</div>
```

#### Mix and Match

```jsx
<div>
  <Text inline={false} color='red'>
    Hello!
  </Text>
  <Text size='small' heavy upper>
    I am text
  </Text>
</div>
```
