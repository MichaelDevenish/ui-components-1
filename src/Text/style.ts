import styled, { css } from 'styled-components'
import { getColor } from '../Color'

export interface TextWrapperProps {
  color?: string | 'subtle'
  inline?: boolean
  upper?: boolean
  weight?: 'normal' | 'heavy'
  size?: 'small' | 'medium' |'large'
  truncated?: boolean
}

export const TextWrapper = styled.span`
  font-family: 'Open Sans', Arial, sans-serif;

  ${(props: TextWrapperProps) => {
    switch (props.color) {
      case 'subtle':
        return `
          color: ${getColor('main-text-light')};
        `
      default:
        return `
          color: ${props.color};
        `
    }
  }}

  ${(props: TextWrapperProps) => {
    switch (props.weight) {
      case 'normal':
        return `
          font-weight: 400;
        `
      case 'heavy':
        return `
          font-weight: 600;
        `
    }
  }}

  ${(props: TextWrapperProps) => props.upper && css`
    text-transform: uppercase;
  `}

  ${(props: TextWrapperProps) => props.truncated && css`
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  `}

  ${(props: TextWrapperProps) => !props.inline && css`
    display: block;
  `}

  ${(props: TextWrapperProps) => {
    switch (props.size) {
      case 'small':
        return `
          font-size: .8125rem;
        `
      case 'medium':
        return `
          font-size: 1rem;
        `
      case 'large':
        return `
          font-size: 1.125rem;
        `
    }
  }}
`
