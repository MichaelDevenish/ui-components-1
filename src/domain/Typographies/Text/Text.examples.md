#### Basic Text

```jsx
  <Text>
    Hello! I am text
  </Text>
```

#### Types

```jsx
  <div>
    <Text isInline={false} type='xsmall'>
      X-Small
    </Text>
    <Text isInline={false} type='small'>
      Small
    </Text>
    <Text isInline={false} type='body'>
      Body (default)
    </Text>
    <Text isInline={false} type='heading'>
      Heading
    </Text>
    <Text isInline={false} type='display'>
      Display
    </Text>
    <Text isInline={false} type='display-large'>
      Display Large
    </Text>
  </div>
```

#### Tags
A tag can be provided to turn any of the Text options into a Heading, or paragraph tag. By default Text is a span.

```jsx
  <div>
    <Text tag='h1' type='display-large'>
      Inspect me! I'm a h1 tag
    </Text>
    <Text tag='p' type='body'>
      Inspect me! I'm a p tag
    </Text>
  </div>
```

#### Weight

```jsx
  <div>
    <Text isInline={false} weight='normal'>
      Normal (default)
    </Text>
    <Text weight='heavy'>
      Heavy
    </Text>
  </div>
```

#### Color

```jsx
const { Variables } = require('../../../common');

  <div>
    <Text color={Variables.Color.b400}>
      Hello! I am blue text
    </Text>
  </div>
```

#### Upper

```jsx
  <Text isUpper>
    Hello! You will see me in all caps
  </Text>
```

#### Inline (false)

```jsx
<div>
  <Text isInline={false}>
    Hello!
  </Text>
  <Text>
    I am text
  </Text>
</div>
```

#### Mix and Match

```jsx
import { Variables } from '@Common';

<div>
  <Text isInline={false} color={Variables.Color.r400}>
    Hello!
  </Text>
  <Text type='small' weight='heavy' isUpper>
    I am text
  </Text>
</div>
```

#### Data attribute

```jsx
  <Text
    dataAttributes={{
      role: 'myRole',
      otherKey: 'myOtherKey',
    }}
  >
    Hello! I am text with data attributes
  </Text>
```

#### Hint (Tooltip)
```jsx
  <Text
    hintComponentProps={{
      hint: 'I can be a string or a JSX Element!'
    }}
  >
    Hover me
  </Text>
```

#### Hint (Popover)
```jsx
  const { Props } = require('@Common');

  <Text
    hintComponentProps={{
      hint: 'I can be a string or a JSX Element!',
      hintType: Props.HintWrapperType.Popover
    }}
  >
    Hover me
  </Text>
```
