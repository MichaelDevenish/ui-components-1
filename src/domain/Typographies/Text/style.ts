import styled, { css } from 'styled-components'
import { getColor } from '../../../common/legacy'

const { n600 } = require('../../../common/sass/variables.scss')

export interface ITextWrapperProps {
  color?: string | 'subtle'
  isInline?: boolean
  isUpper?: boolean
  weight?: 'normal' | 'heavy'
  size?: 'xsmall' | 'small' | 'medium' |'large'
  isTruncated?: boolean
}

export const TextWrapper = styled.span`
  font-family: 'Open Sans', Arial, sans-serif;

  ${(props: ITextWrapperProps) => {
    switch (props.color) {
      case 'subtle':
        return `
          color: ${n600};
        `
      default:
        return `
          color: ${props.color};
        `
    }
  }}

  ${(props: ITextWrapperProps) => {
    switch (props.weight) {
      case 'normal':
        return `
          font-weight: 400;
        `
      case 'heavy':
        return `
          font-weight: 600;
        `
    }
  }}

  ${(props: ITextWrapperProps) => props.isUpper && css`
    text-transform: uppercase;
  `}

  ${(props: ITextWrapperProps) => props.isTruncated && css`
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  `}

  ${(props: ITextWrapperProps) => !props.isInline && css`
    display: block;
  `}

  ${(props: ITextWrapperProps) => {
    switch (props.size) {
      case 'xsmall':
        return `
          font-size: .8125rem;
        `
      case 'small':
        return `
          font-size: .875rem;
        `
      case 'medium':
        return `
          font-size: 1rem;
        `
      case 'large':
        return `
          font-size: 1.125rem;
        `
    }
  }}
`
