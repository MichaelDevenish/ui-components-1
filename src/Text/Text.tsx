import React from 'react'
import { TextWrapper } from './style'
import { withSkeleton, SkeletonComponentProps } from '../Skeleton'

export interface TextProps {
  /** Text to display */
  children: string
  /** Custom class name to use */
  className?: string
  /** Specify if text is a heading */
  isHeading?: boolean
  /** If true, displays the text in uppercase */
  isUpper?: boolean
  /** If true, displays the text with a heavy font weight */
  isHeavy?: boolean
  /** If true, displays the text in a smaller font size */
  isSmall?: boolean
  /** If true, will truncate overflowing text */
  isTruncated?: boolean
  /** If true, will display the text inline */
  isInline?: boolean
  /** Color of the text */
  color?: string
}

export class TextComponent extends React.PureComponent<TextProps> {
  public static defaultProps: Partial<TextProps> = {
    isInline: true
  }

  public render (): JSX.Element {
    const {
      children,
      className,
      isHeading,
      isHeavy,
      isUpper,
      isSmall,
      isTruncated,
      isInline,
      color
    } = this.props

    return (
      <TextWrapper
        color={color}
        isInline={isInline}
        isHeavy={isHeavy}
        isHeading={isHeading}
        isUpper={isUpper}
        isSmall={isSmall}
        isTruncated={isTruncated}
        className={className}
      >
        {children}
      </TextWrapper>
    )
  }
}

export const Text: React.ComponentClass<TextProps & SkeletonComponentProps> = withSkeleton(TextComponent)
