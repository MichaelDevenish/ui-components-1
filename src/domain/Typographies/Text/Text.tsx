import React from 'react'
import { TextWrapper } from './style'
import { withSkeleton, ISkeletonComponentProps } from '../../Skeletons'
import { Variables } from '../../../common'

export interface ITextProps {
  /** Text to display */
  children: string | string[] | number
  /** Custom class name to use */
  className?: string
  /** If true, displays the text in uppercase */
  isUpper?: boolean
  /** Specify the text font weight */
  weight?: 'normal' | 'heavy'
  /** Specify the type of text to use */
  type?: 'xsmall' | 'small' | 'body' | 'heading' | 'display' | 'display-large'
  /** If true, will truncate overflowing text */
  isTruncated?: boolean
  /** If true, will display the text inline */
  isInline?: boolean
  /** Color of the text */
  color?: Variables.Color | 'subtle'
}

export class TextComponent extends React.PureComponent<ITextProps> {
  public static defaultProps: Partial<ITextProps> = {
    isInline: true,
    color: Variables.Color.n800,
    type: 'body',
    weight: 'normal'
  }

  public render (): JSX.Element {
    const {
      children,
      className,
      type,
      weight,
      isUpper,
      isTruncated,
      isInline,
      color
    } = this.props

    return (
      <TextWrapper
        type={type}
        color={color}
        isInline={isInline}
        weight={weight}
        isUpper={isUpper}
        isTruncated={isTruncated}
        className={className}
      >
        {children}
      </TextWrapper>
    )
  }
}

export const Text: React.ComponentClass<ITextProps & ISkeletonComponentProps> = withSkeleton(TextComponent)
