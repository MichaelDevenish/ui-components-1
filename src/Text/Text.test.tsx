import React from 'react'
import { mount } from 'enzyme'

import { Text } from './Text'

describe('<Text />', () => {
  it(`should render an element with text`, () => {
    const wrapper = mount(
      <Text>Hello, world!</Text>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a skeleton element if skeletonOptions are passed to the component`, () => {
    const wrapper = mount(
      <Text
        skeletonOptions={{
          showSkeleton: true,
          shape: 'line'
        }}
      >
        Hello, world!
      </Text>
    )

    expect(wrapper).toMatchSnapshot()
  })
})
