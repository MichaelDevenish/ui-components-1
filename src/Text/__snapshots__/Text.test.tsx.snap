// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should render a skeleton element if skeletonOptions are passed to the component 1`] = `
<class_1
  skeletonOptions={
    Object {
      "shape": "line",
      "showSkeleton": true,
    }
  }
>
  <Skeleton
    className=""
    skeletonOptions={
      Object {
        "shape": "line",
        "showSkeleton": true,
      }
    }
  >
    <span
      className="Skeleton skeleton line skeleton-large"
      style={
        Object {
          "width": undefined,
        }
      }
    >
      â€Œ
    </span>
  </Skeleton>
</class_1>
`;

exports[`<Text /> should render an element with heavy text 1`] = `
.c0 {
  font-weight: 600;
}

<class_1
  isHeavy={true}
>
  <Skeleton
    className=""
    isHeavy={true}
    skeletonOptions={
      Object {
        "shape": "line",
        "showSkeleton": false,
        "size": "large",
      }
    }
  >
    <TextComponent
      isHeavy={true}
      isInline={true}
    >
      <styled.span
        isHeavy={true}
        isInline={true}
      >
        <span
          className="c0"
        >
          Hello, world!
        </span>
      </styled.span>
    </TextComponent>
  </Skeleton>
</class_1>
`;

exports[`<Text /> should render an element with large text 1`] = `
.c0 {
  font-size: 1.125rem;
}

<class_1
  isLarge={true}
>
  <Skeleton
    className=""
    isLarge={true}
    skeletonOptions={
      Object {
        "shape": "line",
        "showSkeleton": false,
        "size": "large",
      }
    }
  >
    <TextComponent
      isInline={true}
      isLarge={true}
    >
      <styled.span
        isInline={true}
        isLarge={true}
      >
        <span
          className="c0"
        >
          Hello, world!
        </span>
      </styled.span>
    </TextComponent>
  </Skeleton>
</class_1>
`;

exports[`<Text /> should render an element with subtle text 1`] = `
.c0 {
  color: hsl(210,13.4%,47.2%);
}

<class_1
  isSubtle={true}
>
  <Skeleton
    className=""
    isSubtle={true}
    skeletonOptions={
      Object {
        "shape": "line",
        "showSkeleton": false,
        "size": "large",
      }
    }
  >
    <TextComponent
      isInline={true}
      isSubtle={true}
    >
      <styled.span
        isInline={true}
        isSubtle={true}
      >
        <span
          className="c0"
        >
          Hello, world!
        </span>
      </styled.span>
    </TextComponent>
  </Skeleton>
</class_1>
`;

exports[`<Text /> should render an element with text 1`] = `
<class_1>
  <Skeleton
    className=""
    skeletonOptions={
      Object {
        "shape": "line",
        "showSkeleton": false,
        "size": "large",
      }
    }
  >
    <TextComponent
      isInline={true}
    >
      <styled.span
        isInline={true}
      >
        <span
          className=""
        >
          Hello, world!
        </span>
      </styled.span>
    </TextComponent>
  </Skeleton>
</class_1>
`;
